pkg/estimator/server.go

```diff
@@ -0,0 +1,302 @@
+
+	// controllerOptions for leader election and client connection
+	controllerOptions *utils.ControllerOptions
+
+	kubeClient kubernetes.Interface
+
+	informerFactory informers.SharedInformerFactory
+	nodeInformer    informerv1.NodeInformer
+	podInformer     informerv1.PodInformer
+	nodeLister      listerv1.NodeLister
+	podLister       listerv1.PodLister
+
+	getPodFunc GetPodsAssignedToNodeFunc
+
+	// default in-tree registry
+	registry  frameworkruntime.Registry
+	framework framework.Framework
+}
+
+func NewEstimatorServer(
+	ctx context.Context,
+	opts *utils.ControllerOptions,
+) (*EstimatorServer, error) {
+	clientConfig, err := utils.LoadsKubeConfig(&opts.ClientConnection)
+	if err != nil {
+		return nil, err
+	}
+	clientConfig.QPS = opts.ClientConnection.QPS
+	clientConfig.Burst = int(opts.ClientConnection.Burst)
+
+	// creating the clientset
+	rootClientBuilder := clientbuilder.SimpleControllerClientBuilder{
+		ClientConfig: clientConfig,
+	}
+	kubeClient := kubernetes.NewForConfigOrDie(rootClientBuilder.ConfigOrDie("clusternet-estimator"))
+	informerFactory := informers.NewSharedInformerFactory(kubeClient, known.DefaultResync)
+
+	broadcaster := record.NewBroadcaster()
+	broadcaster.StartRecordingToSink(&v1core.EventSinkImpl{
+		Interface: kubeClient.CoreV1().Events(""),
+	})
+	utilruntime.Must(appsapi.AddToScheme(scheme.Scheme))
+	utilruntime.Must(clusterapi.AddToScheme(scheme.Scheme))
+	recorder := broadcaster.NewRecorder(scheme.Scheme, corev1.EventSource{Component: "clusternet-estimator"})
+	es := &EstimatorServer{
+		ctx:               ctx,
+		controllerOptions: opts,
+		kubeClient:        kubeClient,
+		informerFactory:   informerFactory,
+		nodeInformer:      informerFactory.Core().V1().Nodes(),
+		podInformer:       informerFactory.Core().V1().Pods(),
+		nodeLister:        informerFactory.Core().V1().Nodes().Lister(),
+		podLister:         informerFactory.Core().V1().Pods().Lister(),
+		registry:          plugins.NewInTreeRegistry(),
+	}
+
+	framework, err := frameworkruntime.NewFramework(es.registry, getDefaultPlugins(),
+		frameworkruntime.WithEventRecorder(recorder),
+		frameworkruntime.WithInformerFactory(informerFactory),
+		frameworkruntime.WithClientSet(kubeClient),
+		frameworkruntime.WithKubeConfig(clientConfig),
+		frameworkruntime.WithParallelism(parallelize.DefaultParallelism),
+		frameworkruntime.WithRunAllFilters(false),
+	)
+	if err != nil {
+		return nil, err
+	}
+	es.framework = framework
+
+	getPodFunc, err := es.BuildGetPodsAssignedToNodeFunc(es.podInformer)
+	if err != nil {
+		return nil, err
+	}
+	es.getPodFunc = getPodFunc
+
+	return es, nil
+}
+
+var _ schedulerapi.EstimatorProvider = &EstimatorServer{}
+
+func (es *EstimatorServer) Run() error {
+	// Start all informers.
+	es.informerFactory.Start(es.ctx.Done())
+	// Wait for all caches to sync before scheduling.
+	es.informerFactory.WaitForCacheSync(es.ctx.Done())
+
+	// if leader election is disabled, so runCommand inline until done.
```

Comment:
Remove leader election part.

This could be added back in cmd.

Reply:
Done